applications/tari_testnet_miner/src/miner.rs

```diff
@@ -0,0 +1,154 @@
+    transaction::*,
+    types::*,
+};
+use tari_crypto::{
+    commitment::HomomorphicCommitmentFactory,
+    keys::{PublicKey as PK, SecretKey},
+    range_proof::RangeProofService,
+};
+use tari_utilities::byte_array::ByteArray;
+
+/// TestNet Miner
+pub struct Miner {
+    // current working block
+    block: Option<Block>,
+    // prev block's solved difficulty, TODO check this
+    difficulty: Difficulty,
+    // current consensus rules
+    rules: ConsensusRules,
+}
+
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum MinerError {
+    // Could not construct the coinbase utxo and kernel for the block
+    CoinbaseError,
+    // No block provided to mine
+    MissingBlock,
+}
+
+impl Miner {
+    /// Create a new empty miner
+    pub fn new(rules: ConsensusRules) -> Miner {
+        Miner {
+            block: None,
+            difficulty: Difficulty::min(),
+            rules,
+        }
+    }
+
+    /// Add a block to be processed.  This function also add the coinbase tx
+    pub fn add_block(&mut self, mut block: Block, coinbase_key: PrivateKey) -> Result<(), MinerError> {
+        let mut rng = rand::OsRng::new().unwrap();
+        let coinbase_value =
+            block.calculate_coinbase_and_fees(self.rules.emission_schedule().block_reward(block.header.height));
+        let new_commitment = COMMITMENT_FACTORY.commit(&coinbase_key, &coinbase_value.into());
+        let rr = PROVER
+            .construct_proof(&coinbase_key, coinbase_value.into())
+            .map_err(|_| MinerError::CoinbaseError)?;
+        let coinbase = TransactionOutput {
+            commitment: new_commitment,
+            features: OutputFeatures::create_coinbase(block.header.height, &self.rules),
+            proof: RangeProof::from_bytes(&rr).unwrap(),
+        };
+        let excess = COMMITMENT_FACTORY.commit(&coinbase_key, &(MicroTari(0)).into());
+        let nonce = PrivateKey::random(&mut rng);
+        let challenge = Miner::get_challenge(&PublicKey::from_secret_key(&nonce));
+        let sig = Signature::sign(coinbase_key, nonce, &challenge).map_err(|_| MinerError::CoinbaseError)?;
+        let kernel = TransactionKernel {
+            features: KernelFeatures::empty(),
+            fee: 0.into(),
+            lock_height: 0,
+            excess,
+            excess_sig: sig,
+            meta_info: None,
+            linked_kernel: None,
+        };
+        block.body.outputs.push(coinbase);
+        block.body.kernels.push(kernel);
+        self.block = Some(block);
+        Ok(())
+    }
+
+    /// This function will mine the nonce and fill out the header.
+    // Todo convert into futures with multi threading for this function
+    pub fn mine(&mut self, old_header: BlockHeader) -> Result<(), MinerError> {
+        if self.block.is_none() {
+            return Err(MinerError::MissingBlock);
+        }
+        let interval = self.block.as_ref().unwrap().header.timestamp.timestamp() - old_header.timestamp.timestamp();
```

Comment:
TODO: We'll bring a community-led difficulty adjustment algo into the core consensus rules at some point. This is ok for now.